### ISSUE TITLE
VictoriaMetrics returns a "hole" in a metric where Prometheus does not

### ISSUE THREAD

**Issue by g7r** [VictoriaMetrics/VictoriaMetrics#1773]
**Describe the bug**
I'm working with a setup where Prometheus scrapes metrics from its targets and stores them into VictoriaMetrics cluster. Missing a single scrape shouldn't be a big deal. It happens from time to time. But I see holes in my Grafana graphs which directly queries VictoriaMetrics cluster's vmselects.

**To Reproduce**
Start VictoriaMetrics inside Docker container:
```
$ docker run -it --rm -p 8428:8428 victoriametrics/victoria-metrics
```

Import metrics into VictoriaMetrics:
```
$ curl --data-binary $'test_metric 27111895040 1635750605.174\ntest_metric 27130007552 1635750639.113\ntest_metric 27146936320 1635750669.335\ntest_metric 27175477248 1635750704.104\ntest_metric 27193053184 1635750730.921\ntest_metric 27216207872 1635750762.536\ntest_metric 27229102080 1635750778.684\ntest_metric 27248865280 1635750810.269\ntest_metric 27276951552 1635750843.259\ntest_metric 27301896192 1635750881.298\ntest_metric 27308892160 1635750897.341\ntest_metric 27340759040 1635750946.925\ntest_metric 27357949952 1635750970.745\ntest_metric 27378688000 1635751003.326\ntest_metric 27398369280 1635751031.302\ntest_metric 27417493504 1635751062.361\ntest_metric 27438039040 1635751088.992\ntest_metric 27452444672 1635751108.574\ntest_metric 27478491136 1635751147.131\ntest_metric 27499462656 1635751174.055\n' -X POST 'http://127.0.0.1:8428/api/v1/import/prometheus'
```

Make a query:
```
$ curl 'http://127.0.0.1:8428/api/v1/query_range?query=test_metric&start=1635750930&end=1635750950&step=2'
```

**Expected behavior**
The result should include all points in specified range with 2s interval. Prometheus behaves so on the same input data.

**Actual behavior**
The result doesn't include 5 points with timestamps in range [1635750938, 1635750946].

**Version**
`victoria-metrics-20211022-171312-tags-v1.68.0-0-gf64f62692`

**Comment by valyala** [comment]
Prometheus fills gaps in time series if they don't exceed 5 minutes. See [these docs](https://prometheus.io/docs/prometheus/latest/querying/basics/#staleness) for details. VictoriaMetrics automatically detects the median interval between samples per each time series on the requested time range and fills gaps only if the interval between raw samples doesn't exceed 2x of the detected median interval.

The following solutions exist in order to fill bigger gaps in VictoriaMetrics:
* To wrap the queried metric name into [last_over_time](https://docs.victoriametrics.com/MetricsQL.html#last_over_time) function with the lookbehind window, which exceeds the maximum gap, which you want to fill. For example, `last_over_time(test_metric[5m])` would fill gaps with the length of up to 5 minutes like Prometheus does.
* To run VictoriaMetrics with `-search.minStalenessInterval=5m` command-line flag. In this case VictoriaMetrics will fill all the gaps smaller than 5 minutes even if it detects much smaller interval between samples in a time series.

**Comment by g7r** [comment]
It is quite a pain to operate on mixed VictoriaMetrics/Prometheus stack due to a number such discrepancies. What concerns me most is that these discrepancies aren't due to some architectural differences. They exists because of deliberate decisions made by VictoriaMetrics developers.
There definitely are ways to implement Prometheus functionality without breaking compatibility with it. I'm OK with having different syntax for different behaviour. I'm also OK with having different non-default behaviour enabled by flags. But the way the things are now should force people to think twice before adopting VictoriaMetrics.
I don't have a strong opinion that Prometheus is doing all right and VictoriaMetrics is doing wrong or vice versa. My point is that there shouldn't be such discrepancies, no matter who's right.
Otherwise VictoriaMetrics is a great tool which is a pleasure to work with.

**Comment by valyala** [comment]
The following troubleshooting guide may help debugging and resolving issues with gaps on VictoriaMetrics graphs - https://docs.victoriametrics.com/Troubleshooting.html#unexpected-query-results

**Comment by valyala** [comment]
It has been appeared that the issue with gaps on graphs for metrics scraped from `cadvisor` has been appeared starting from [v1.68.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.68.0). `cadvisor` exports metrics with incorrect timestamps - see [this issue](https://github.com/google/cadvisor/issues/2526). These timestamps were ignored by VictoriaMetrics and `vmagent` before `v1.68.0`. These timestamps started to be used instead of local scrape timestamps starting from `v1.68.0`, and this broke staleness detection algorithm in VictoriaMetrics, which, in turn, resulted in gaps on graphs for metrics obtained from `cadvisor`. See [this](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4697#issuecomment-1654614799) and [this](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4697#issuecomment-1656540535) comments for details.

This issue has been fixed in the commit e3ef3df938fb8f6afe8694ad9308e22e5f0b8ba8 , which will be included in the next release.

**Comment by valyala** [comment]
Closing this issue as a duplicate for https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4697

**Comment by valyala** [comment]
FYI, the issue with gaps on graphs built from `cadvisor` metrics, has been fixed starting from [v1.93.0](https://docs.victoriametrics.com/CHANGELOG.html).
